### ISSUE TITLE
ROPE operator

### ISSUE THREAD

**Issue by prajaramanTT** [tenstorrent/tt-metal#15941]
Like ALiBi, AFAICT TTNN demos uses Torch to precompute the ROPE tensors. Which on GGML
has to fallback onto the CPU. ROPE is is worst offender in terms of number of operators. > 1000
of them in a single forward pass of TinyLLaMA. Causing major slowdown due to to unnecessary
data copy and (un-)tilizing.

**Comment by uaydonat** [comment]
@avoraTT our new ROPE computes the ROPE tensors on device, correct?

**Comment by avoraTT** [comment]
> [@avoraTT](https://github.com/avoraTT) our new ROPE computes the ROPE tensors on device, correct?


Yes, the sin/cos matrices are selected on device using the `ttnn.embedding` op.

**Comment by avoraTT** [comment]
> Like ALiBi, AFAICT TTNN demos uses Torch to precompute the ROPE tensors. Which on GGML has to fallback onto the CPU. ROPE is is worst offender in terms of number of operators. > 1000 of them in a single forward pass of TinyLLaMA. Causing major slowdown due to to unnecessary data copy and (un-)tilizing.

Are you referring to the use of `ttnn.experimental.rotary_embedding_llama` or `ttnn.experimental.rotary_embedding`?

When using `ttnn.experimental.rotary_embedding_llama` in decode mode, the ROPE tensors are precomputed once for all positions (until 128k in the case of Llama3), and then sent to device (DRAM Interleaved). Then when decoding, the `ttnn.embedding` op is used with some `position_ids` to select the respective tensors. Ie, there should be no host overhead during decode.

**Comment by uaydonat** [comment]
@prajaramanTT when was the original issue filed? Which op are they using?

As Ammar mentions above, we recently made improvements to this op. Are they using the new `rotay_embedding_llama` ?

**Comment by marty1885** [comment]
GGML issues are basically all mine - some are filed via TT because Saboo asked me for a list of missing ops. I will try. But I'm occupied rn. CNY is coming up then FOSDEM.

The original comment was made moths ago

**Comment by marty1885** [comment]
@uaydonat How do I use `rotary_embedding`? I am looking at the docs and confused. How do I generate the cache?

```
        Args:
            input_tensor (ttnn.Tensor): the input tensor.
            cod_cache (ttnn.Tensor): the Cosine Cache tensor.
            sin_cache (ttnn.Tensor): the Sine Cache tensor.
            token_index (int, optional): Defaults to `None`.
```

**Comment by avoraTT** [comment]
@marty1885 a good starting point on how to use `ttnn.experimental.rotary_embedding_llama` would be to check out the [tech report](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/LLMs/llms.md#22-rope) on LLM bring up. As an overview, our `rotary_embedding_llama` implementation is based off of eqation 34 from the [RoFormer paper](https://arxiv.org/pdf/2104.09864). The transformation matrix is generated using the `get_rot_transformation_mat` function [here](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/llama3/tt/llama_common.py#L141).

For prefill mode, the cos/sin cache are `[1, 1, seq_len, dhead]` tensors with the precomputed frequences. For an example of how we generate the cos/sin cache, please take a look at our unit test [here](https://github.com/tenstorrent/tt-metal/blob/main/tests/tt_eager/python_api_testing/unit_testing/misc/test_rotary_embedding_llama.py#L245-L249). Essentially, we call `compute_gather_cos_sin` with our specified `dhead` and `position_ids`.


For decode mode, the cos/sin cache are generated similarly, but the shape is `[1, batch, 1, dhead]`. One thing to note, is that in the Llama use case, we precompute all the cos/sin matrices for all position ids (up to 128k) and then store them on DRAM. Then at inference time, we use the `ttnn.emebedding` op combined with `position_ids` to select the correct freqs for each user. See and example implementation [here](https://github.com/tenstorrent/tt-metal/blob/main/models/demos/llama3/tt/llama_rope.py), with it's associated test file [here](https://github.com/tenstorrent/tt-metal/blob/main/tests/tt_eager/python_api_testing/unit_testing/misc/test_rotary_embedding_llama.py).


Please let me know if you have any questions!

**Comment by marty1885** [comment]
@avoraTT I finally got time working on this subject again. Thanks for the links - I'm still confused even after reading the paper. What is the transformation matrix in `rotary_embedding_llama` doing and how do I specify token idxs in the API? At this point I feel I know enough about the the hardware and RoPE itself to write my own kernel. But having trouble mapping it back to TTNN APIs. Do you have documentation on what exactly  `rotary_embedding_llama` is doing?

**Comment by avoraTT** [comment]
@marty1885 apologies for the delay, I've written up an overview of how the prefill `rotary_embedding_llama` is implemented/used. Please let me know if you have any questions, or need to know more about the decode version (it is a similar concept with some slight changes)!



[RoPE Documentation.pdf](https://github.com/user-attachments/files/19454319/RoPE.Documentation.pdf)

**Comment by marty1885** [comment]
Closing as I implemented my own in my free time. Ref
https://github.com/marty1885/llama.cpp/blob/metalium-support/ggml/src/ggml-metalium/kernels/rope_normal_compute.cpp
